src/uShip.Logging/LogBuilders/FluentLogDataBuilder.cs

```diff
@@ -82,6 +84,104 @@ public IFluentLoggerWriter Exception(Exception exception)
                 return this;
             }
 
+            /// <summary>
+            /// Reflectively add non-complex properties (exceptions like DateTime and DateTimeOffset) from an object to the Data dictionary to be included in the logger's message.
+            /// </summary>
+            /// <param name="prefix">Prefix that will be prepended to the property key names in the data dictionary.</param>
+            /// <param name="value">Object/class that we are logging properties from.</param>
+            /// <returns></returns>
+            public IFluentLoggerWriter Data(string prefix, object value)
+            {
+                MapDataFromObject(prefix, value, true);
+                return this;
+            }
+
+            /// <summary>
+            /// Short-hand version of <see cref="Data(string,object)"/ of logging data from an object.
+            /// </summary>
+            /// <param name="value">Object's whose type will be used as the prefix to logging key name for all its properties.</param>
+            /// <returns></returns>
+            public IFluentLoggerWriter Data(object value)
+            {
+                var name = value.GetType().Name + '_';
+                return Data(name, value);
+            }
+
+            /// <summary>
+            /// Recursive method that adds properties of an object to the Data dictionary to be included in the logger's message.
+            /// </summary>
+            /// <param name="name"></param>
+            /// <param name="value"></param>
+            /// <param name="topLevel"></param>
+            private void MapDataFromObject(string name, object value, bool topLevel = false)
+            {
+                var type = value.GetType();        
+                if (topLevel)
+                {                    
+                    var props = type.GetProperties();
+                    //hard limit to 30 for now
+                    //TODO: Extend this to be dynamic.
+                    foreach (var prop in props.Take(30))
+                    {
+                        var propertyPrefixedName = name + prop.Name;
+                        //this seems really odd, but it is how you peel the properties's values off of the topLevel object. 
+                        var val = prop.GetValue(value, null);
+                        MapDataFromObject(propertyPrefixedName, val);
+                    }
+                    //we infer that this is definitely not a primitive and exit early. 
+                    return;
+                }
+                                
+                if (type.IsPrimitive
```

Comment:
This is a great point. As long as we rely on the topLevel construct/flow, we don't necessarily need to check out types here.